UnknownServerException: java.io.IOException: Broken pipe will be thrown and cause job fail when connection is closed by server

### Search before asking

- [x] I searched in the [issues](https://github.com/alibaba/fluss/issues) and found nothing similar.


### Fluss version

main (development)

### Please describe the bug 🐞

When run this job, sometimes it may throw `com.alibaba.fluss.exception.UnknownServerException: java.io.IOException: Broken pipe` and cause the Flink job fail. The detail exception is 
```
Caused by: java.util.concurrent.CompletionException: com.alibaba.fluss.exception.UnknownServerException: java.io.IOException: Broken pipe
	at sun.nio.ch.FileDispatcherImpl.write0(Native Method)
	at sun.nio.ch.SocketDispatcher.write(SocketDispatcher.java:47)
	at sun.nio.ch.IOUtil.writeFromNativeBuffer(IOUtil.java:95)
	at sun.nio.ch.IOUtil.write(IOUtil.java:53)
	at sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:492)
	at com.alibaba.fluss.shaded.netty4.io.netty.channel.socket.nio.NioSocketChannel.doWrite(NioSocketChannel.java:415)
	at com.alibaba.fluss.shaded.netty4.io.netty.channel.AbstractChannel$AbstractUnsafe.flush0(AbstractChannel.java:931)
	at com.alibaba.fluss.shaded.netty4.io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.flush0(AbstractNioChannel.java:355)
	at com.alibaba.fluss.shaded.netty4.io.netty.channel.AbstractChannel$AbstractUnsafe.flush(AbstractChannel.java:895)
	at com.alibaba.fluss.shaded.netty4.io.netty.channel.DefaultChannelPipeline$HeadContext.flush(DefaultChannelPipeline.java:1372)
	at com.alibaba.fluss.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:921)
	at com.alibaba.fluss.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:907)
	at com.alibaba.fluss.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:893)
	at com.alibaba.fluss.shaded.netty4.io.netty.channel.ChannelDuplexHandler.flush(ChannelDuplexHandler.java:127)
	at com.alibaba.fluss.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:923)
	at com.alibaba.fluss.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:941)
	at com.alibaba.fluss.shaded.netty4.io.netty.channel.AbstractChannelHandlerContext$WriteTask.run(AbstractChannelHandlerContext.java:1247)
	at com.alibaba.fluss.shaded.netty4.io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
	at com.alibaba.fluss.shaded.netty4.io.netty.util.concurrent.Abstr
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:607)
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:591)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990)
	at com.alibaba.fluss.client.table.writer.AbstractTableWriter.lambda$send$0(AbstractTableWriter.java:77)
	at com.alibaba.fluss.client.write.WriteBatch.lambda$completeFutureAndFireCallbacks$0(WriteBatch.java:173)
```

### Solution

The exception happens when the server close the netty channel, while the client still send request via the channel almost at the same time.. After I add some debug, I found the cause..
1: At first, the server close the channel due to channel idle
```
2025-02-27 04:05:27,237 WARN  com.alibaba.fluss.rpc.netty.server.NettyServerHandler        [] - Connection /ip:60678 is idle, closing...
```
2: Almost at some time, the client log the following exception(note the logging is not in main branch, it's added for my debug purpose)
```
 ERROR com.alibaba.fluss.rpc.netty.client.ServerConnection [] - Fail to writeAndFlush for local address /ip:60678...
java.io.IOException: Broken pipe
```
First of all, the `UnknownServerException` really mislead me to make me feel like it's a exception thrown by server, but turns out not. We will wrap any unknow exception receive from server or just happens in network as `UnknownServerException`.

Then, the cause is sometime the server close the netty channel, and then send message  [in here](https://github.com/alibaba/fluss/blob/main/fluss-rpc/src/main/java/com/alibaba/fluss/rpc/netty/client/ServerConnection.java#L323), the IOException(broken pipeline) is thrown.. 

I'd like to propose to wrap the IOException as `NetworkException` which is retrable to enable client to re-send the request again.


### Are you willing to submit a PR?

- [x] I'm willing to submit a PR!

--- a/fluss-rpc/src/main/java/com/alibaba/fluss/rpc/netty/client/ServerConnection.java
+++ b/fluss-rpc/src/main/java/com/alibaba/fluss/rpc/netty/client/ServerConnection.java
@@ -327,8 +327,15 @@ final class ServerConnection {
                                         if (!future.isSuccess()) {
                                             connectionMetricGroup.updateMetricsAfterGetResponse(
                                                     apiKey, inflight.requestStartTime, 0);
-                                            inflight.responseFuture.completeExceptionally(
-                                                    future.cause());
+                                            Throwable cause = future.cause();
+                                            if (cause instanceof IOException) {
+                                                // when server close the channel, the cause will be
+                                                // IOException, if the cause is IOException, we wrap
+                                                // it as retryable NetworkException to retry to
+                                                // connect
+                                                cause = new NetworkException(cause);
+                                            }
+                                            inflight.responseFuture.completeExceptionally(cause);
                                             inflightRequests.remove(inflight.requestId);
                                         }
                                     });

